content/post/2020-11-18-testthat-utility-belt/index.Rmd

```diff
@@ -0,0 +1,106 @@
+* R scripts under `tests/testthat/` whose name start with `helper-` are loaded with `devtools::load_all()` so they are available for both tests and interactive debugging. Now, it also means their behavior does not different from R scripts under `R/` so you might put them in that directory too as recommended in testthat docs. However, it also means they are installed with the package which might (slightly!) increase its size, and that they are with the rest of your package code which might [put you off](https://community.rstudio.com/t/why-are-tests-testthat-helper-files-discouraged-in-testthat/85253).
+
+To summarize,
+
+| File                           | Run before tests | Loaded via `load_all()` | Installed with the package[^installed] | Testable[^testable] |
+|--------------------------------|------------------|-------------------------|----------------------------|----------|
+| tests/testthat/setup-.\*.R     | ✔️               | \-                      | \-                         | \-       |
+| tests/testthat/helper-.\*.R    | ✔️               | ✔️                      | \-                         | \-       |
+| R/any-name.R                   | ✔️               | ✔️                      | ✔️                         | ✔️       |
+| tests/testthat/anything-else.R | \-               | \-                      | \-                         | \-       |
+
+`tests/testthat/helper-.*.R` are [no longer recommended in testthat](https://testthat.r-lib.org/reference/test_dir.html#special-files) but they are still supported. :relieved:
+
+[^installed]: Installed with the package, and [run at package installation time](https://github.com/r-lib/testthat/issues/1206#issuecomment-713519962).
+[^testable]: Yes, you could [test the code supporting your tests](https://github.com/r-lib/testthat/issues/1206#issuecomment-713583205).
+
+In practice,
+
+* In `tests/testthat/setup.R` you might do something like loading a package that helps your unit testing like `{vcr}`, `{httptest}` or `{presser}` if you're [testing an API client](https://books.ropensci.org/http-testing/).
+* In a helper like `tests/testthat/helper.R` or `R/test-helpers.R` you might define variables and functions that you'll use throughout your tests, even [custom skippers](https://testthat.r-lib.org/articles/skipping.html#helpers). To choose between the two locations, refer to the table above and your own needs and preferences. Note that if someone wanted to study testthat "utility belts" à la [Bob Rudis](https://rud.is/b/2018/04/08/dissecting-r-package-utility-belts/), they would probably only identify helper files like `tests/testthat/helper.R`.
+
+You'll notice testthat no longer recommends having a file with code to be run after tests... So how do you clean up after tests?
```

Comment:
Can we just say teardown files (as well)?

Reply:
well one should no longer use them?

Reply:
Right. Never mind then.